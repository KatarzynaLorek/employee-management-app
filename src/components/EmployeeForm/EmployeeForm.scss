.form {
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: 700px;
  width: 100vw;
  max-height: 700px;
  height: 100vh;
  background-color: white;
  padding: 10px 10px;

  &__wrapper {
    display: flex;
    align-items: center;
    justify-content: center;

    &--inner {
      @extend .form__wrapper;
      flex-direction: column;
      max-width: 600px;
      width: 100%;
    }

    &--outer {
      @extend .form__wrapper;
      position: fixed;
      top: 0;
      left: 0;
      width: 100vw;
      height: 100vh;
      background-color: rgba(0, 0, 0, 0.2);
    }
  }

  &__title {
    display: block;
    padding-top: 30px;
    height: 100px;
  }

  &__button {
    transition: color 0.2s linear;
    height: 40px;
    background-color: #fff;
    cursor: pointer;

    &--close {
      @extend .form__button;
      align-self: flex-end;
      width: 40px;
      margin: 5px 5px;
      border: none;
      border-radius: 20px;
      font-size: 25px;
      line-height: 40px;

      &:hover {
        color: crimson;
        transition: color 0.2s linear;
      }
    }

    &--submit {
      @extend .form__button;
      width: 100px;
      margin-top: 70px;
      border: 2px solid #77b28c;
      border-radius: 10px;
      font-size: 1.6rem;
      line-height: 40px;
      text-transform: uppercase;
      transition: background-color 0.2s linear;

      &:hover {
        background-color: #77b28c;
        color: white;
        font-weight: 700;
      }
    }
  }

  &__label {
    height: 30px;
    margin: 20px 0 10px 0;
    font-size: 2rem;
    align-self: flex-start;
    line-height: 30px;
  }

  &__input {
    height: 40px;
    width: 100%;
    padding-left: 10px;
    font-size: 1.6rem;
  }

  &__error {
    align-self: flex-start;
    padding: 5px 0 0 10px;
    color: red;
    font-weight: 700;
  }
}

@media (max-width: 1250px) {
  .form {
    &__wrapper--outer {
      padding-top: 220px;
    }
  }
}

@media (max-height: 800px) {
  .form {
    &__wrapper--outer {
      overflow: scroll;
      scrollbar-width: 0;
    }
  }
}
