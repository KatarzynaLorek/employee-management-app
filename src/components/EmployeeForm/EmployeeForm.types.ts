import { IResponseObject } from '../../types/responses';

export enum EmployeeFormVariant {
  add = 'add',
  update = 'update',
}
export interface IEmployeeForm {
  handleClose: () => void;
  id: string;
  submitAction: (employee: IResponseObject) => void;
  employeeData?: IResponseObject;
}
