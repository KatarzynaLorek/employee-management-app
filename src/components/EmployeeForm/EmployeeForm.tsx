import React from 'react';
import { Formik, Field, Form, FormikHelpers } from 'formik';
import './EmployeeForm.scss';
import { IResponseObject } from '../../types/responses';

interface IForm {
  handleClose: () => void;
  employeeData?: IResponseObject | null;
  submitAction: (employee: IResponseObject) => Promise<any>;
  newID?: string;
}

const EmployeeForm = ({ handleClose, submitAction, employeeData, newID }: IForm): JSX.Element => {
  const initialValues: IResponseObject = {
    id: employeeData?.id || newID,
    firstName: employeeData?.firstName || '',
    lastName: employeeData?.lastName || '',
    department: employeeData?.department || 'accountancy',
    position: employeeData?.position || 'junior',
  };

  const handleSubmit = (
    newValues: IResponseObject,
    setSubmitting: (boolean: boolean) => void,
  ): void => {
    submitAction(newValues);
    setSubmitting(false);
    handleClose();
  };

  const validateTextInput = (values: IResponseObject) => {
    const errors: { firstName?: string; lastName?: string } = {};
    if (!values.firstName) {
      errors.firstName = 'Required';
    }
    if (!values.lastName) {
      errors.lastName = 'Required';
    }
    return errors;
  };

  return (
    <div className="form__wrapper form__wrapper--outer">
      <div className="form">
        <button
          aria-label="closeButton"
          className="form__button form__button--close"
          onClick={handleClose}
        >
          <i className="fas fa-times"></i>
        </button>
        <h1 className="form__title">
          {employeeData ? 'Update Employee Data' : 'Add new employee'}
        </h1>
        <Formik
          validate={validateTextInput}
          initialValues={initialValues}
          onSubmit={(
            values: IResponseObject,
            { setSubmitting }: FormikHelpers<IResponseObject>,
          ) => {
            handleSubmit(values, setSubmitting);
          }}
        >
          {({ errors, touched }) => (
            <Form
              name={employeeData ? 'updateForm' : 'addForm'}
              className="form__wrapper form__wrapper--inner"
            >
              <label className="form__label" htmlFor="firstName">
                First Name
              </label>
              <Field
                autoComplete="off"
                className="form__input"
                id="firstName"
                name="firstName"
                placeholder="John"
              />
              {errors.firstName && touched.firstName && (
                <div className="form__error">{errors.firstName}</div>
              )}
              <label className="form__label" htmlFor="lastName">
                Last Name
              </label>
              <Field
                autoComplete="off"
                className="form__input"
                id="lastName"
                name="lastName"
                placeholder="Doe"
              />
              {errors.lastName && touched.lastName && (
                <div className="form__error">{errors.lastName}</div>
              )}
              <label className="form__label" htmlFor="department">
                Department
              </label>
              <Field className="form__input" as="select" id="department" name="department">
                <option value="accountancy">Accountancy</option>
                <option value="IT">IT</option>
                <option value="marketing">Marketing</option>
                <option value="sales">Sales</option>
              </Field>
              <label className="form__label" htmlFor="position">
                Position
              </label>
              <Field as="select" className="form__input" id="position" name="position" type="text">
                <option value="junior">Junior</option>
                <option value="regular">Regular</option>
                <option value="senior">Senior</option>
                <option value="manager">Manager</option>
              </Field>
              <button className="form__button form__button--submit" type="submit">
                {employeeData ? 'Update' : 'Submit'}
              </button>
            </Form>
          )}
        </Formik>
      </div>
    </div>
  );
};

export default EmployeeForm;
