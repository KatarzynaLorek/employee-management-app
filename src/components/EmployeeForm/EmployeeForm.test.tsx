import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import EmployeeForm from './EmployeeForm';

const renderAddForm = () => {
  const mockCloseFn = jest.fn();

  const utils = render(<EmployeeForm formType="add" handleClose={mockCloseFn} />);

  const form = utils.getByRole('form');

  const firstNameInput = utils.getByRole('textbox', { name: 'First Name' });
  const lastNameInput = utils.getByRole('textbox', { name: 'Last Name' });
  const positionInput = utils.getByRole('combobox', { name: 'Position' });
  const departmentInput = utils.getByRole('combobox', {
    name: 'Department',
  });

  const submitButton = utils.getByRole('button', {
    name: /submit/i,
  });
  const closeButton = utils.getByRole('button', { name: 'closeButton' });
  return {
    ...utils,
    mockCloseFn,
    form,
    submitButton,
    closeButton,
    firstNameInput,
    lastNameInput,
    positionInput,
    departmentInput,
  };
};

describe('EmployeeForm component', () => {
  it('renders properly', () => {
    const { form } = renderAddForm();
    expect(form).toBeInTheDocument();
  });

  it('renders different types of form', () => {
    const { submitButton, rerender, getByRole } = renderAddForm();
    expect(submitButton).toBeInTheDocument();

    rerender(<EmployeeForm formType="update" handleClose={() => jest.fn()} />);
    expect(getByRole('button', { name: /update/i })).toBeInTheDocument();
  });

  it('runs close function on click', () => {
    const { closeButton, mockCloseFn } = renderAddForm();
    fireEvent.click(closeButton);
    expect(mockCloseFn.mock.calls.length).toBe(1);
  });

  it('displays proper values', async () => {
    const { firstNameInput, lastNameInput, positionInput, departmentInput } = renderAddForm();

    const mockName = 'mockName';
    await waitFor(() => {
      fireEvent.change(firstNameInput, { target: { value: mockName } });
      fireEvent.change(lastNameInput, { target: { value: mockName } });
    });
    expect(firstNameInput).toHaveValue(mockName);
    expect(lastNameInput).toHaveValue(mockName);

    await waitFor(() => {
      fireEvent.change(departmentInput, { target: { value: 'IT' } });
    });
    expect(departmentInput).toHaveValue('IT');

    await waitFor(() => {
      fireEvent.change(positionInput, {
        target: {
          value: 'senior',
        },
      });
    });
    expect(positionInput).toHaveValue('senior');
  });

  it('validates values on submit', async () => {
    const { firstNameInput, lastNameInput, submitButton, container } = renderAddForm();

    await waitFor(() => {
      fireEvent.click(submitButton);
    });
    expect(container).toHaveTextContent('Required');

    const mockName = 'mockName';
    await waitFor(() => {
      fireEvent.change(firstNameInput, { target: { value: mockName } });
    });
    expect(container).toHaveTextContent('Required');

    await waitFor(() => {
      fireEvent.change(lastNameInput, { target: { value: mockName } });
    });
    expect(container).not.toHaveTextContent('Required');
  });
});
