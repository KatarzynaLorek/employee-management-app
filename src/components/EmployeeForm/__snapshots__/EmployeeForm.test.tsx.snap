// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeForm component renders properly an add form 1`] = `
<body>
  <div>
    <div
      class="form__wrapper--outer"
    >
      <div
        class="form"
      >
        <button
          aria-label="closeButton"
          class="form__button--close"
        >
          <i
            class="fas fa-times"
          />
        </button>
        <h1
          class="form__title"
        >
          Add new employee
        </h1>
        <form
          action="#"
          class="form__wrapper--inner"
          name="addForm"
        >
          <label
            class="form__label"
            for="firstName"
          >
            First Name
          </label>
          <input
            autocomplete="off"
            class="form__input"
            id="firstName"
            name="firstName"
            placeholder="John"
            value=""
          />
          <label
            class="form__label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            autocomplete="off"
            class="form__input"
            id="lastName"
            name="lastName"
            placeholder="Doe"
            value=""
          />
          <label
            class="form__label"
            for="department"
          >
            Department
          </label>
          <select
            class="form__input"
            id="department"
            name="department"
          >
            <option
              value="accountancy"
            >
              Accountancy
            </option>
            <option
              value="IT"
            >
              IT
            </option>
            <option
              value="marketing"
            >
              Marketing
            </option>
            <option
              value="sales"
            >
              Sales
            </option>
          </select>
          <label
            class="form__label"
            for="position"
          >
            Position
          </label>
          <select
            class="form__input"
            id="position"
            name="position"
            type="text"
          >
            <option
              value="junior"
            >
              Junior
            </option>
            <option
              value="regular"
            >
              Regular
            </option>
            <option
              value="senior"
            >
              Senior
            </option>
            <option
              value="manager"
            >
              Manager
            </option>
          </select>
          <button
            class=" form__button--submit"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</body>
`;

exports[`EmployeeForm component renders properly an update form 1`] = `
<body>
  <div>
    <div
      class="form__wrapper--outer"
    >
      <div
        class="form"
      >
        <button
          aria-label="closeButton"
          class="form__button--close"
        >
          <i
            class="fas fa-times"
          />
        </button>
        <h1
          class="form__title"
        >
          Update Employee Data
        </h1>
        <form
          action="#"
          class="form__wrapper--inner"
          name="updateForm"
        >
          <label
            class="form__label"
            for="firstName"
          >
            First Name
          </label>
          <input
            autocomplete="off"
            class="form__input"
            id="firstName"
            name="firstName"
            placeholder="John"
            value="mockFirstName"
          />
          <label
            class="form__label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            autocomplete="off"
            class="form__input"
            id="lastName"
            name="lastName"
            placeholder="Doe"
            value="mockLastname"
          />
          <label
            class="form__label"
            for="department"
          >
            Department
          </label>
          <select
            class="form__input"
            id="department"
            name="department"
          >
            <option
              value="accountancy"
            >
              Accountancy
            </option>
            <option
              value="IT"
            >
              IT
            </option>
            <option
              value="marketing"
            >
              Marketing
            </option>
            <option
              value="sales"
            >
              Sales
            </option>
          </select>
          <label
            class="form__label"
            for="position"
          >
            Position
          </label>
          <select
            class="form__input"
            id="position"
            name="position"
            type="text"
          >
            <option
              value="junior"
            >
              Junior
            </option>
            <option
              value="regular"
            >
              Regular
            </option>
            <option
              value="senior"
            >
              Senior
            </option>
            <option
              value="manager"
            >
              Manager
            </option>
          </select>
          <button
            class=" form__button--submit"
            type="submit"
          >
            Update
          </button>
        </form>
      </div>
    </div>
  </div>
</body>
`;
